cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -796,5 +796,12 @@ func ValidateCertValidity(cfg *kubeadm.ClusterConfiguration) []error {
 			errors.Errorf("caCertificateValidityPeriod: the value %v is more than the recommended default for CA certificate expiration: %v",
 				cfg.CACertificateValidityPeriod.Duration, constants.CACertificateValidityPeriod))
 	}
+	if cfg.CertificateValidityPeriod != nil && cfg.CACertificateValidityPeriod != nil {
+		if cfg.CertificateValidityPeriod.Duration > cfg.CACertificateValidityPeriod.Duration {
```

Comment:
reading online we should probably not be doing that check.
i also haven't seen it in core k8s PKI.

>> Summary: you cannot rely on validity period nesting, and you should not try to enforce nesting when validating certificates.

https://security.stackexchange.com/questions/120469/can-a-ssl-certificate-have-longer-validity-period-than-its-parent-in-x-509-chain

my 2c: if the user decided to have a shorter lived CA it would be their problem.
would we face bugs in such conditions?

Reply:
We didn't allow to change the validity period before, so we may don't test that case.
If the user set the CA expiration time too short, although CertificateValidityPeriod certificate have set up a long time, the k8s clusters will become unavailable when CA is invalid.

```sh
# kubeadm certs check-expiration
[check-expiration] Reading configuration from the cluster...
[check-expiration] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
[check-expiration] Error reading configuration from the Cluster. Falling back to default configuration

W0805 10:41:37.054107   10150 certs.go:475] WARNING: could not validate bounds for certificate CA: the certificate has expired: NotBefore: 2024-08-05 10:33:29 +0000 UTC, NotAfter: 2024-08-05 10:41:29 +0000 UTC
W0805 10:41:37.054841   10150 certs.go:475] WARNING: could not validate bounds for certificate etcd CA: the certificate has expired: NotBefore: 2024-08-05 10:33:29 +0000 UTC, NotAfter: 2024-08-05 10:41:29 +0000 UTC
W0805 10:41:37.055918   10150 certs.go:475] WARNING: could not validate bounds for certificate front-proxy CA: the certificate has expired: NotBefore: 2024-08-05 10:33:29 +0000 UTC, NotAfter: 2024-08-05 10:41:29 +0000 UTC
CERTIFICATE                EXPIRES                  RESIDUAL TIME   CERTIFICATE AUTHORITY   EXTERNALLY MANAGED
admin.conf                 Aug 05, 2025 10:38 UTC   364d            ca                      no      
apiserver                  Aug 05, 2025 10:38 UTC   364d            ca                      no      
apiserver-etcd-client      Aug 05, 2025 10:38 UTC   364d            etcd-ca                 no      
apiserver-kubelet-client   Aug 05, 2025 10:38 UTC   364d            ca                      no      
controller-manager.conf    Aug 05, 2025 10:38 UTC   364d            ca                      no      
etcd-healthcheck-client    Aug 05, 2025 10:38 UTC   364d            etcd-ca                 no      
etcd-peer                  Aug 05, 2025 10:38 UTC   364d            etcd-ca                 no      
etcd-server                Aug 05, 2025 10:38 UTC   364d            etcd-ca                 no      
front-proxy-client         Aug 05, 2025 10:38 UTC   364d            front-proxy-ca          no      
scheduler.conf             Aug 05, 2025 10:38 UTC   364d            ca                      no      
super-admin.conf           Aug 05, 2025 10:38 UTC   364d            ca                      no      

CERTIFICATE AUTHORITY   EXPIRES                  RESIDUAL TIME   EXTERNALLY MANAGED
ca                      Aug 05, 2024 10:41 UTC   <invalid>       no      
etcd-ca                 Aug 05, 2024 10:41 UTC   <invalid>       no      
front-proxy-ca          Aug 05, 2024 10:41 UTC   <invalid>       no 

# kubectl get node
Unable to connect to the server: tls: failed to verify certificate: x509: certificate has expired or is not yet valid: current time 2024-08-05T10:41:42Z is after 2024-08-05T10:41:29Z
```

Given that we have this configuration option open, do we really not need to alert users to possible misconfigurations?

Reply:
> A warning might help users spot potential configuration problems, and if they are intentional, they can just ignore the warning. We really shouldn't treat it as an error, but a warning seems ok.

ok, i don't think i mind the warning and the overall period nesting topic is debatable.
also it would be surprising if users have the CA period to be less than the leaf validity.

Reply:
+1 to add the warning.

> also it would be surprising if users have the CA period to be less than the leaf period.

Can we avoid this by adding a validation for `CertificateValidityPeriod` should less than `CACertificateValidityPeriod`?